pkg/controllers/user/pipeline/controller/pipelineexecution/registrycertsyncer.go

```diff
@@ -0,0 +1,180 @@
+	"github.com/rancher/rancher/pkg/ticker"
+	"github.com/rancher/types/apis/core/v1"
+	"github.com/rancher/types/apis/project.cattle.io/v3"
+	"github.com/sirupsen/logrus"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/util/cert"
+	"time"
+)
+
+// This Syncer is responsible for certificates rotation for internal registries
+
+const (
+	checkCertRotateInterval = 12 * time.Hour
+	//rotateThreshold set the rotation threshold so it will rotate at 90% of the total lifetime of the certificates.
+	rotateThreshold = 0.9
+	maxRetry        = 60
+)
+
+type RegistryCertSyncer struct {
+	clusterName string
+
+	pods                    v1.PodInterface
+	podLister               v1.PodLister
+	secrets                 v1.SecretInterface
+	secretLister            v1.SecretLister
+	managementSecretLister  v1.SecretLister
+	namespaceLister         v1.NamespaceLister
+	pipelineExecutionLister v3.PipelineExecutionLister
+	pipelineSettingLister   v3.PipelineSettingLister
+}
+
+func (s *RegistryCertSyncer) sync(ctx context.Context, syncInterval time.Duration) {
+	for range ticker.Context(ctx, checkCertRotateInterval) {
+		s.checkAndRotateCerts(ctx)
+	}
+}
+
+func (s *RegistryCertSyncer) checkAndRotateCerts(ctx context.Context) {
+	labelsSearchSet := labels.Set{utils.PipelineNamespaceLabel: "true"}
+	namespaces, err := s.namespaceLister.List("", labels.SelectorFromSet(labelsSearchSet))
+	if err != nil {
+		logrus.Error(err)
+		return
+	}
+	for _, ns := range namespaces {
+		secret, err := s.secretLister.Get(ns.Name, utils.RegistryCrtSecretName)
+		if apierrors.IsNotFound(err) {
+			continue
+		} else if err != nil {
+			logrus.Error(err)
+			return
+		}
+		if !s.shouldRotate(secret.Data[utils.RegistryCrt]) {
+			continue
+		}
+		projectID := getProjectID(ns)
+		if projectID != "" {
+			go func() {
+				if err := s.rotateCerts(ctx, projectID); err != nil {
+					logrus.Error(err)
+				}
+			}()
+		}
+	}
+
+}
+
+func (s *RegistryCertSyncer) shouldRotate(certPEM []byte) bool {
+	currentCerts, err := cert.ParseCertsPEM(certPEM)
+	if err != nil || len(currentCerts) < 1 {
+		return true
+	}
+	currentCert := currentCerts[0]
+	totalDuration := currentCert.NotAfter.Sub(currentCert.NotBefore)
+	thresholdTime := currentCert.NotBefore.Add(time.Duration(float64(totalDuration) * rotateThreshold))
+	if time.Now().Before(thresholdTime) {
+		return false
+	}
+	return true
+}
+
+func (s *RegistryCertSyncer) rotateCerts(ctx context.Context, projectID string) error {
+	logrus.Debugf("rotating registry certs for %s project", projectID)
+
+	clusterID, projectID := ref.Parse(projectID)
+	//do certificate rotation when no pipeline execution is running
+	startTime := time.Now()
+	set := labels.Set(map[string]string{utils.PipelineFinishLabel: "false"})
+	cancelCtx, cancelFunc := context.WithCancel(ctx)
+	defer cancelFunc()
+	for range ticker.Context(cancelCtx, time.Minute) {
```

Comment:
what if project gets removed? does this loop ever ends? We should either fetch the project before rotating and checking for its existance/deletion timestamp, or pass the cancel context on project removal

Reply:
On a second thought, if the project is removed and the project namespace is gone, we will get empty array here and break the loop. So I think we don't need the project check here. And we have a retry limit to avoid nonstop loop